File: PrettyPasteboard/PPPasteboardObserver.m
//
//  PPPasteboardObserver.m
//  PrettyPasteboard
//
//  Created by AlexDenisov on 6/1/14.
//  Copyright (c) 2014 Railsware. All rights reserved.
//

#import "PPPasteboardObserver.h"

static const char *kQueueLabel = "org.okolodev.PrettyPasteboard";
static const useconds_t kPollInterval = 250000;

typedef
NS_ENUM(unsigned char, PPPasteboardObserverState)
{
    PPPasteboardObserverStateDisabled,
    PPPasteboardObserverStateEnabled
};

@implementation PPPasteboardObserver
{
    NSMutableSet *_subscribers;
    dispatch_queue_t _serialQueue;
    PPPasteboardObserverState _state;
    NSInteger _changeCount;
}

#pragma mark - LifeCycle

- (instancetype)init
{
    self = [super init];
    _serialQueue = dispatch_queue_create(kQueueLabel, NULL);
    _changeCount = -1;
    _subscribers = [NSMutableSet new];
    return self;
}

- (void)dealloc
{
    [self stopObserving];
    [self removeSubscribers];
}

#pragma mark - Observing

- (void)startObserving
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self changeState:PPPasteboardObserverStateEnabled];
        [self observerLoop];
    });
}

- (void)stopObserving
{
    [self changeState:PPPasteboardObserverStateDisabled];
}

- (void)changeState:(PPPasteboardObserverState)state
{
    dispatch_sync(_serialQueue, ^{
        _state = state;
    });
}

- (BOOL)isEnabled
{
    return _state == PPPasteboardObserverStateEnabled;
}

- (void)observerLoop
{
    while ([self isEnabled]) {
        if (_changeCount != self.pasteboard.changeCount) {
            _changeCount = self.pasteboard.changeCount;
            [self pasteboardContentChanged];
        }
        usleep(kPollInterval);
    }
}

- (NSPasteboard *)pasteboard
{
    if (!_pasteboard) {
        _pasteboard = [NSPasteboard generalPasteboard];
    }
    
    return _pasteboard;
}

- (void)pasteboardContentChanged
{
    for (id<PPPasteboardObserverSubscriber> subscriber in _subscribers) {
        [subscriber pasteboardChanged:self.pasteboard];
    }
}

#pragma mark - Subscribers

- (void)addSubscriber:(id<PPPasteboardObserverSubscriber>)subscriber
{
    [_subscribers addObject:subscriber];
}

- (void)removeSubscriber:(id<PPPasteboardObserverSubscriber>)subscriber
{
    [_subscribers removeObject:subscriber];
}

- (void)removeSubscribers
{
    [_subscribers removeAllObjects];
}

@end

File: PrettyPasteboard/PPPasteboardObserver.swift
//
//  PPPasteboardObserver.swift
//  PrettyPasteboard
//
//  Created by AlexDenisov on 6/3/14.
//  Copyright (c) 2014 Railsware. All rights reserved.
//

import Foundation
import Cocoa

enum PPPasteboardObserverState {
    case PPPasteboardObserverStateDisabled
    case PPPasteboardObserverStateEnabled
    case PPPasteboardObserverStatePaused
}

class PPPasteboardObserver : NSObject {
    var pasteboard : NSPasteboard = NSPasteboard.generalPasteboard()
    
    var subscribers : NSMutableSet = NSMutableSet()
    var serialQueue : dispatch_queue_t = dispatch_queue_create("org.okolodev.PrettyPasteboard", nulDev)
    var changeCount : Int = -1
    var state : PPPasteboardObserverState = PPPasteboardObserverState.PPPasteboardObserverStateDisabled
    
    init() {
        
    }
    
    deinit {
        self.stopObserving()
        self.removeSubscribers()
    }
    
    // Observing
    
    func startObserving() {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), { () -> Void in
            self.changeState(PPPasteboardObserverState.PPPasteboardObserverStateEnabled)
            self.observerLoop()
        });
    }
    
    func stopObserving() {
        self.changeState(PPPasteboardObserverState.PPPasteboardObserverStateDisabled)
    }

    func pauseObserving() {
        self.changeState(PPPasteboardObserverState.PPPasteboardObserverStatePaused)
    }
    
    func continueObserving() {
        if (self.state == PPPasteboardObserverState.PPPasteboardObserverStatePaused) {
            self.changeCount = self.pasteboard.changeCount;
            self.state = PPPasteboardObserverState.PPPasteboardObserverStateEnabled
        }
    }
    
    func observerLoop() {
        while self.isEnabled() {
            usleep(250000)
            var countEquals = self.changeCount == self.pasteboard.changeCount
            if countEquals {
                continue
            }
            
            self.changeCount = self.pasteboard.changeCount
            self.pasteboardContentChanged()
        }
    }
    
    func pasteboardContentChanged() {
        self.pauseObserving()
        for anySubscriber : AnyObject in self.subscribers {
            if let subscriber = anySubscriber as? protocol<PPPasteboardObserverSubscriber> {
                subscriber.pasteboardChanged(self.pasteboard)
            }
        }
        self.continueObserving()
    }
    
    func changeState(newState: PPPasteboardObserverState) {
        dispatch_sync(self.serialQueue, { () -> Void in
            self.state = newState;
        });
    }
    
    func isEnabled() -> Bool {
        return self.state == PPPasteboardObserverState.PPPasteboardObserverStateEnabled;
    }
    
    // Subscribers
    
    func addSubscriber(subscriber: PPPasteboardObserverSubscriber) {
        self.subscribers.addObject(subscriber)
    }
    
    func removeSubscribers() {
        self.subscribers.removeAllObjects()
    }

}
